[Runtime][Win32] Fix fatalError() backtraces.

al45tair opened the pull request:
We were calling `SymInitialize()` multiple times, which is wrong, which was making the `SymbolInfo::lookup()` call fail.  Also, we weren't fetching the module names, so we should do that too.

rdar://130992923

al45tair committed:
[Runtime][Win32] Fix fatalError() backtraces.

We were calling `SymInitialize()` multiple times, which is wrong, which
was making the `SymbolInfo::lookup()` call fail.  Also, we weren't
fetching the module names, so we should do that too.

rdar://130992923

al45tair commented:
@swift-ci Please smoke test

grynspan commented on stdlib/public/runtime/SymbolInfo.cpp:
```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+
+  if (bRet) {
+    return SymbolInfo((const void *)package.si.Address,
+                      ::strdup(package.si.Name),
```
I was trying to avoid the heap allocation here, FWIW, since we might be crashing due to allocation failure.

al45tair replied:
FWIW, we do heap allocation all over the shop in the surrounding code, so I'm presently unconcerned about that. If this was actually for a crash handler, as opposed to `fatalError()`, I'd be more worried, obviously.

grynspan replied:
Fair enough.

grynspan commented on stdlib/public/runtime/SymbolInfo.h:
```diff
@@ -43,14 +43,74 @@ namespace swift {
 struct SymbolInfo {
 private:
 #if defined(_WIN32) && !defined(__CYGWIN__)
-  SYMBOL_INFO_PACKAGE _package;
```
What's your thinking in not storing the whole `SYMBOL_INFO_PACKAGE` here? Less transient memory?

al45tair replied:
Partly that. Also, when `SymFromAddr()` fails, I wanted to still give the module name and base address, as well as indicating the address we were looking at (but obviously no symbol).

al45tair replied:
It's swings and roundabouts though, because of the `::malloc()`s.

grynspan commented on stdlib/public/runtime/SymbolInfo.h:
```diff
@@ -43,14 +43,74 @@ namespace swift {
 
+#if defined(_WIN32) && !defined(__CYGWIN__)
+  SymbolInfo(const SymbolInfo &other) {
+    _symbolName = nullptr;
```
You don't want to duplicate these?

al45tair replied:
Ah, the duplication is done by the `*this = other` line just below. The reason for setting these to `nullptr` was to make sure we don't fall over by calling `free()` with uninitialised pointer values.

grynspan commented on stdlib/public/runtime/ImageInspectionCOFF.cpp:
```diff
@@ -54,8 +55,7 @@ void _swift_win32_withDbgHelpLibrary(
     // If we have not previously initialized the Debug Help library, do so now.
-    bool isDbgHelpInitialized = false;
-    if (dbgHelpHandle) {
+    if (dbgHelpHandle && !isDbgHelpInitialized) {
```
So, it's more complicated than this. We are _supposed_ to call `SymInitialize()` multiple times, but follow them up with calls to `SymCleanup()`. Which is probably what's missing.

(At one point, Swift Testing was going to hook into `_swift_win32_withDbgHelpLibrary()`, but it doesn't do the backtrace work we thought we needed, so it's become moot.)

al45tair replied:
Why would we want to call `SymInitialize()` and `SymCleanup()` over and over? That's going to be expensive, I would imagine, as it's going to go off looking for symbols and things.

grynspan replied:
Because this library is thread-unsafe to call in the general case, and at least in theory calling SymCleanup() means client code that calls into it is less likely to stomp.

al45tair replied:
The documentation for `SymInitialize()` is unhelpfully vague, but my reading of it is that it should be OK to call `SymInitialize()` multiple times with _different_ process handles — that is, I think you can have multiple sessions active simultaneously. The API is still thread-unsafe, so needs mutual exclusion, but I don't think we need to `SymInitialize()` and `SymCleanup()` each time.

grynspan replied:
It's admittedly been a while since I touched this code, and I've paged the dbghelp knowledge out of my head by now.

compnerd commented on stdlib/public/runtime/SymbolInfo.cpp:
```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+
+    if (!GetModuleInformation(GetCurrentProcess(), hModule,
+                              &mi, sizeof(mi))) {
+      ::memset(&mi, 0, sizeof(mi));
```
I'd use `ZeroMemory` to avoid the Android-style `memset` bug.

```suggestion
      ZeroMemory(&mi, sizeof(mi));
```

al45tair replied:
I have no preference either way. `ZeroMemory()` makes sense given that this is Windows code.

compnerd commented on stdlib/public/runtime/SymbolInfo.cpp:
```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+
+    WCHAR wszBuffer[256];
+    LPWSTR pwszFileName = wszBuffer;
+    DWORD dwCapacity = sizeof(wszBuffer) / sizeof(WCHAR);
```
```suggestion
    DWORD dwCapacity = sizeof(wszBuffer) / sizeof(*wszBuffer);
```

compnerd commented on stdlib/public/runtime/SymbolInfo.h:
```diff
@@ -43,14 +43,74 @@ namespace swift {
+    if (_moduleFileName)
+      ::free((void *)_moduleFileName);
+    if (_symbolName)
+      ::free((void *)_symbolName);
```
`free(nullptr)` is well defined, I'm not sure that there is value in the explicit check.

al45tair replied:
I'm aware that `::free(nullptr)` is well-defined; I didn't have the tests originally, then added them to match the ones in the copy assignment operator. I think you're right that we should remove them in the destructor.

(Nobody pointed out that I hadn't initialised them in the `SymbolInfo()` constructor; _that_ is clearly a bug, so I'll fix that and make this change separately.)

compnerd commented on stdlib/public/runtime/SymbolInfo.h:
```diff
@@ -43,14 +43,74 @@ namespace swift {
+      if (_moduleFileName)
+        ::free((void *)_moduleFileName);
+      if (_symbolName)
+        ::free((void *)_symbolName);
```
Likewise

al45tair replied:
The reason I added it here is that I chose to implement the constructor using `*this = other;`, and I didn't like that calling `::free(nullptr)` twice every time on construction.

al45tair replied:
I'll change it to not use `*this = other;` and to use a private member function instead, I think. That avoids my motivation for the tests.

al45tair commented on stdlib/public/runtime/SymbolInfo.cpp:
```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+
+// Get the filename and base of the module that contains the specified
+// address.  N.B. This returns a malloc()ed copy of the filename in the
+// Win32ModuleInfo struct; you are responsible for freeing that.
```
Yes. We want the caller to own the memory so it can pass it straight into the `SymbolInfo` structure without having to copy it again.

al45tair commented on stdlib/public/runtime/SymbolInfo.cpp:
```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+  const char *name;
+  const void *base;
+
+  Win32ModuleInfo(const char *n, const void *b) : name(n), base(b) {}
```
Doing that would be wrong, unless we also added a `SymbolInfo` constructor that took an rvalue-reference to a `Win32ModuleInfo`. Which we could do, but then `Win32ModuleInfo` would need to be public, which I don't really like. It's really just being used to allow the `moduleInfoFromAddress()` function to return two values.

al45tair replied:
I'd rather use brace initialisation, I think, in that case. The point of this is just to carry the allocated name out of the function; it isn't supposed to be used outside of that.

al45tair commented on stdlib/public/runtime/SymbolInfo.cpp:
```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+// Win32ModuleInfo struct; you are responsible for freeing that.
+static Win32ModuleInfo moduleInfoFromAddress(const void *address) {
+  HMODULE hModule;
+  if (GetModuleHandleEx(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS
+                        | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT,
```
I'm not sure what that buys us. I guess it makes the `GetModuleHandleEx()` call fit on one line, but we gain another line or two for the `dwFlags` variable, and it's also potentially going to move the flags away from the call site.

al45tair committed:
[Runtime][Win32] Tweak `SymbolInfo` slightly based on review comments.

Prefer `ZeroMemory()` to `::memset()` here.

Use `sizeof(*wszBuffer)` instead of `sizeof(WCHAR)`, just in case.

Don't use `*this=other`, because that motivates tests around `::free()`,
but instead pull the shared code out into some private functions.

Also, fix `SymbolInfo()` to initialize the pointer members.

rdar://130992923

al45tair commented:
@swift-ci Please smoke test

compnerd commented on stdlib/public/runtime/SymbolInfo.cpp:
```diff
@@ -59,42 +60,119 @@ const char *SymbolInfo::getSymbolName() const {
+    }
+
+    WCHAR wszBuffer[256];
+    LPWSTR pwszFileName = wszBuffer;
```
Can we use `std::unique_ptr<WSTR>`? That would simplify the exit path.

grynspan replied:
The default destructor uses global operator delete I think, which is something you'd really want to avoid.

grynspan replied:
How about `std:array`?

compnerd replied:
`std::unique_ptr` takes an optional custom deleter, which can use `free`. It is just a tad more verbose to write:

```c++
std::unique_ptr<WSTR, decltype(std::free)> buffer{nullptr, std::free};
buffer.reset(std::realloc(buffer.release(), size << 1));
```

al45tair committed:
[Runtime][Win32] Further tweaks to SymbolInfo.cpp and Errors.cpp.

Fix an unused variable warning in `Errors.cpp`.

Use brace initialization syntax in `SymbolInfo.cpp` rather than using
a constructor call.

rdar://130992923

al45tair commented:
@swift-ci Please smoke test
